BCnEnc.Net/Shared/AsyncOperation.cs

```diff
@@ -0,0 +1,52 @@
+using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace BCnEncoder.Shared
+{
+	public class AsyncOperation<T> : IDisposable
+	{
+		private CancellationTokenSource tokenSource;
+		private Task<T> task;
+
+		public T Result => task.Result;
+
+		public Task<T> Start(Func<CancellationToken, T> func)
+		{
+			if (task != null)
+			{
+				throw new InvalidOperationException("This instance is already running an operation or finished it.");
+			}
+
+			var token = CreateTokenSource().Token;
+			return task = Task.Factory.StartNew(() => func(token), token);
+		}
+
+		public void Cancel()
+		{
+			tokenSource.Cancel();
+		}
+
+		/// <summary>
+		/// Allows awaiting the asynchronous operation.
+		/// </summary>
+		/// <returns>The awaiter for the running operation.</returns>
+		/// <remarks>The operation should be started before awaiting it.</remarks>
+		public TaskAwaiter<T>? GetAwaiter()
+		{
+			return task?.GetAwaiter();
+		}
+
```

Comment:
The Dispose method would need to be called. Or the developer let's it get garbage collected.
I could think about using IAsyncDisposable, which provides an async compatible Dispose method. With this, one could wrap the await into a using block like this:
```
await using (var operation = DecodeAsync(...))
{
    // Do stuff with operation after it was completed, but before it gets disposed
}
```

I don't like this code, since it adds bulk, but if you're ok with this approach, then I can probably add that. Otherwise only having IDisposable, will need to have it called separately.

Reply:
Is there a way to make the api methods only return a Task\<T\> then? Having to use using blocks like that doesn't seem very user friendly, when the Image\<Rgba\> is already IDisposable.

Reply:
I implemented the AsyncOperation to also introduce Cancel support. If that is nothing for you (which would make the async API useless btw) then I can remove it.

Reply:
What about letting the user pass a cancellation token as a parameter if they need cancellation? Using the async without cancellation is not really useless either, since you can launch the task from the UI thread without having to wrap it in a Task.Run call.

I don't really know the standard practice for this but a lot of the async methods I've seen take the token as a parameter.

Reply:
Hmm, the methods are already packed with parameters (and be it only the raw methods), I wouldn't like to add them as parameters.

So the cancellationToken would be passed either via options or BcDecoder ctor. Both of those approaches would make BcDecoder not thread-safe anymore. You would need to create a BcDecoder everytime you want to decode something. That also doesn't feel right either.

So I came up with the AsyncOperation class, which is awaitable and offers a Cancel method.

Also no matter how you do it, without the AsyncAPI you either block the UI thread by not wrapping the call in a Task.Run, or you wrap it in a Task.Run and cannot cancel the operation, which makes UI experience increasingly worse. So having native support for async calls in a library is mostly to allow for cancellation via the library.

But I agree, the normal approach would be to pass a CancellationToken to the methods. I just didn't like this here as explained above. I guess you just tell me what I should do. But I would like to keep the AsyncOperation, which wraps both cancellation support and the running task into one nice class.

Reply:
I tried using the async api to see how it feels, but it says `TaskAwaiter<Image<Rgba>>? does not contain a definition for IsCompleted`.

This is how I called it:
```
using FileStream fs = File.OpenRead(@"../../../testImages/test_decompress_bc1.ktx");
BcDecoder decoder = new BcDecoder();
using var operation = decoder.DecodeAsync(fs);
using var img = await operation;
```

I think it would be better to make the api work the same as Microsoft stuff. Eg. [here](https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.readasync). Have a version of the async function without CancellationToken parameter and one version with it. 

I also found [this](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken.none?view=net-5.0#System_Threading_CancellationToken_None). Should this be used instead of `default(CancellationToken)` or are they the same?

Reply:
Ah, I made the mistake to return a nullable TaskAwaiter<T>. That is stupid.
Your code will work with my next commit. GetAwaiter of the AsyncOperation will now throw an exception if no task is currently running. Which should never happen for the consuming developers, since the operation is started before being passed back to the developer.

Reply:
I understand the reasoning behind it but I also feel like this library's api should "feel like" any other async api. The abundance of parameters does not feel like a solid enough reason to force the user to use this package in a different way from the norm. I do appreciate the effort you put in to come up with the AsyncOperation class but I think using it as a consumer of the api is just too different.

Reply:
Hmm, I see. I accept that POV. Since this is your library I will obviously do as you feel is best. Then I will remove the AsyncOperation and allow the user to pass in a CancellationToken as a normal parameter.